ENH: sparse: extend COO arrays to n-dimensions

anushkasuyal opened the pull request:
This addresses the enhancement proposed (at gh-21169) to extend the COO sparse array support in SciPy to n-dimensions. This PR implements the basic methods for COO arrays, along with arithmetic operations, and matrix-vector multiplication.

I have also added new tests for all the supported methods, and enhanced and extended the existing 1D and 2D array tests, while removing previously asserted behaviour (e.g., an error for >2-D) which no longer applies.

anushkasuyal committed:
DOC: Typeset code objects in gcrotmk documentation

Corrected the typesetting of code objects in the gcrotmk documentation. Ensured that 'A' in the parameter description for 'M' and 'callback(xk)' in the description of 'callback' are properly formatted as code.

anushkasuyal committed:
Merge remote-tracking branch 'upstream/main'

anushkasuyal committed:
Merge remote-tracking branch 'upstream/main'

anushkasuyal committed:
coo_todense_nd added and working

anushkasuyal committed:
WIP - test_coo.py, _add_sparse() causing errors

anushkasuyal committed:
latest

anushkasuyal committed:
nd in working state similar to 3d

anushkasuyal committed:
coo_matvec_3d working

anushkasuyal committed:
coo_matvec_3d working, test fixed

anushkasuyal committed:
coo_matvec_nd working in coo.h

anushkasuyal committed:
added parametrization to tests

anushkasuyal committed:
coo arrays extended to ndim for basic operations and matvec multiplication

anushkasuyal committed:
removed (unused, added at an earlier stage) coo_todense3d and coo_matvec_3d from coo.h

anushkasuyal committed:
lint check changes

anushkasuyal committed:
space removed

anushkasuyal committed:
type comparison using is instead of ==

lucascolley commented on scipy/sparse/_coo.py:
```diff
@@ -286,13 +290,17 @@ def toarray(self, order=None, out=None):
-        M, N = self._shape_as_2d
-        coo_todense(M, N, self.nnz, self.row, self.col, self.data,
-                    B.ravel('A'), fortran)
+        if self.ndim<3:
```
```suggestion
        if self.ndim < 3:
```

lucascolley commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +576,90 @@ def _add_dense(self, other):
-        M, N = self._shape_as_2d
-        coo_todense(M, N, self.nnz, self.row, self.col, self.data,
-                    result.ravel('A'), fortran)
+        if self.ndim<3:
```
```suggestion
        if self.ndim < 3:
```

lucascolley commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +576,90 @@ def _add_dense(self, other):
+    
+
+    def _matmul_vector(self, other):
+        if self.ndim<3:
```
```suggestion
        if self.ndim < 3:
```

lucascolley commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +576,90 @@ def _add_dense(self, other):
+    def _matmul_vector(self, other):
+        if self.ndim<3:
+            result_shape = self.shape[0] if self.ndim > 1 else 1
+            result = np.zeros(result_shape,
+                            dtype=upcast_char(self.dtype.char, other.dtype.char))
```
```suggestion
            result = np.zeros(result_shape,
                              dtype=upcast_char(self.dtype.char, other.dtype.char))
```

lucascolley commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +576,90 @@ def _add_dense(self, other):
+            return result
+        
+        else: # if dim >= 3
+            result = np.zeros(np.prod(self.shape[:-1]),
+                            dtype=upcast_char(self.dtype.char, other.dtype.char))
```
```suggestion
            result = np.zeros(np.prod(self.shape[:-1]),
                              dtype=upcast_char(self.dtype.char, other.dtype.char))
```

lucascolley commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +576,90 @@ def _add_dense(self, other):
+            result = self._matmul_vector(other.ravel())
+            return result.reshape(M, 1)
+        
+        raise NotImplementedError
+        ("n-D matrix-matrix multiplication not implemented for ndim>2")
```
I would keep the `(` on the first line if you are to write it like this, like you have done on line 628. Consider also:
```suggestion
        msg = "n-D matrix-matrix multiplication not implemented for ndim>2"
        raise NotImplementedError(msg)
```

You could consider using this pattern elsewhere to reduce the amount of multi-line strings (but not a big deal either way).

lucascolley commented on scipy/sparse/_sputils.py:
```diff
@@ -241,21 +241,33 @@ def isintlike(x) -> bool:
-        return False
-    for d in x:
-        if not isintlike(d):
+    if ndim<3:
```
```suggestion
    if ndim < 3:
```

lucascolley commented on scipy/sparse/_sputils.py:
```diff
@@ -331,14 +349,25 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
-            raise ValueError('shape must be a 2-tuple of positive integers')
-        if any(d < 0 for d in new_shape):
-            raise ValueError("'shape' elements cannot be negative")
+        if len(new_shape)<3:
```
```suggestion
        if len(new_shape) < 3:
```

lucascolley commented on scipy/sparse/_sputils.py:
```diff
@@ -331,14 +349,25 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+        if len(new_shape)<3:
+            if allow_1d:
+                if len(new_shape) not in (1, 2):
+                    raise ValueError('shape must be a 1- or 2-tuple of positive '
+                                    'integers')
```
```suggestion
                    raise ValueError('shape must be a 1- or 2-tuple of positive '
                                     'integers')
```

lucascolley commented on scipy/sparse/_sputils.py:
```diff
@@ -331,14 +349,25 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+                    raise ValueError("'shape' elements cannot be negative")
+            else:
+                if allow_1d:
+                    raise ValueError('shape must be a 1- or 2-tuple of positive '
+                                    'integers')
```
```suggestion
                    raise ValueError('shape must be a 1- or 2-tuple of positive '
                                     'integers')
```

lucascolley commented on scipy/sparse/_sputils.py:
```diff
@@ -362,8 +391,14 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
 
-    if len(new_shape) != 2 and not (allow_1d and len(new_shape) == 1):
-        raise ValueError('matrix shape must be two-dimensional')
+    if len(new_shape)<3:
```
```suggestion
    if len(new_shape) < 3:
```

lucascolley commented on scipy/sparse/_sputils.py:
```diff
@@ -362,8 +391,14 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+    if len(new_shape)<3:
+            if len(new_shape) != 2 and not (allow_1d and len(new_shape) == 1):
+                raise ValueError('matrix shape must be two-dimensional')
+    elif len(new_shape)>=3:
```
```suggestion
    elif len(new_shape) >= 3:
```

lucascolley commented on scipy/sparse/tests/test_coo.py:
```diff
@@ -102,43 +164,98 @@ def test_reshape():
+        arr2d.reshape((1,3))
+
+
+@pytest.mark.parametrize(('shape', 'new_shape'), [((4,9,6,5), (3,6,15,4)), ((4,9,6,5),
+                                            (36,30)), ((4,9,6,5), (1080,)), ((4,9,6,5),
+                                                                    (2,3,2,2,3,5,3))])
```
```suggestion
@pytest.mark.parametrize(('shape', 'new_shape'), [((4,9,6,5), (3,6,15,4)),
                                                  ((4,9,6,5), (36,30)),
                                                  ((4,9,6,5), (1080,)),
                                                  ((4,9,6,5), (2,3,2,2,3,5,3)),])
```

(the trailing `,` is conventional, so that the line does not need to be changed to add to the list in future)

lucascolley commented on scipy/sparse/tests/test_coo.py:
```diff
@@ -203,15 +320,44 @@ def test_2d_to_1d_resize(arg: int):
+    # 4d case
+    arr4d = coo_array(([2, 3, 7], ([1, 0, 1], [0, 2, 0], [1, 2, 1], [1, 0, 1])))
+    assert arr4d.nnz == 3
+    assert_equal(arr4d.toarray(), 
+                 np.array([[[[0, 0], [0, 0], [0, 0]], [[0, 0], [0, 0], [0, 0]],
+                          [[0, 0], [0, 0], [3, 0]]], [[[0, 0], [0, 9], [0, 0]],
+                          [[0, 0], [0, 0], [0, 0]], [[0, 0], [0, 0], [0, 0]]]]))
```
consider writing like this to make the structure of the array clear. (Unfortunately, this still goes 1 character over our general 88 char limit, so using `  # noqa: E501` will tell the linter to allow us to break the rule here. A similar pattern could be used below.

```suggestion
    expected = np.array(  # noqa: E501
        [[[[0, 0], [0, 0], [0, 0]], [[0, 0], [0, 0], [0, 0]], [[0, 0], [0, 0], [3, 0]]],
         [[[0, 0], [0, 9], [0, 0]], [[0, 0], [0, 0], [0, 0]], [[0, 0], [0, 0], [0, 0]]]]
    )
    assert_equal(arr4d.toarray(), expected)
```

lucascolley reviewed (commented):
(cursory style review)

Code quality looks really good overall, thanks @anushkasuyal ! Just some style nitpicks below.

The doc build failure in CI is unrelated to this PR (gh-21195), however the `ValueError: low is out of bounds for int32` COO test failures on some of the Windows jobs look real.

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_sputils.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_sputils.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_sputils.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_sputils.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_sputils.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/_sputils.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/tests/test_coo.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal committed:
Update scipy/sparse/tests/test_coo.py

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

anushkasuyal commented:
Thanks for all the suggestions @lucascolley ! I will look into the `ValueError`. I suspect it may possibly have to do with some overflow in `np.ravel_multi_index`.

dschult commented on scipy/sparse/sparsetools/coo.h:
```diff
@@ -112,6 +112,43 @@ void coo_todense(const I n_row,
 
 
+template <class I, class T>
+void coo_todense_nd(const I D[],        // array of dimension sizes
```
The term "dimension sizes" is called "shape" now (though maybe some older c++ code used that term).
I also prefer the variable name `shape` over `D` for readability.  I'm also not sure what the name `A` refers to in this function. The code uses it as the coords array, so maybe name it something to indicate that.

Similar name suggestions for other functions (and these are just suggestions. If you think D is more readable just say so).

anushkasuyal replied:
Changing `D` to `shape` sounds like a reasonable choice! As for `A`, I agree that the name isn't indicating anything. Since it is the coords array, for uniformity with `Ax` for data array (which I took from `coo_todense`), I could rename `A` to `Ac`. If that's not as sensible either, then we could change `Ac` to `A_coords` and `Ax` to `A_data` (or something similar). The commit at 5823bbe used `Ac`, but if you recommend `A_coords`, then I can make the changes!

dschult replied:
Some of the other functions use `Ai` and `Aj` names for the coords in 2D. Perhaps this could be `Aijk`.
I don't think `Ac` is better than `A`. But `Acoo` or `A_coords` would be fine.  The `Ax` is used in **lots** of functions in `sparsetools`, so I think you should keep that as is.

anushkasuyal replied:
Changed to `Aijk` at 8f4075f.

dschult reviewed (commented):
Thanks very much for this!
I have a number of nit-picky comments below and a few bigger questions/suggestions here.

From my quick reading, it looks like `_matmul_dispatch` will return shape (M,1) instead of `self.shape[:-1]` when `other` is a column. So are we losing all the dimensions but the last two?

The tests that use `randn` to construct random numpy arrays should either be made not random, or be given a seed to make them deterministic. Its hard to debug a test that only fails occasionally. Also, I think the tests with RandomState should (by numpy recommendation) be changed to use the random generator interface -- even though the keyword they are provided to is named `random_state`. :}

And finally, I would really like to know how much slower it would be to use numpy tools instead of the `coo_todense_nd` C++ code. Something like:
```python
    result = np.zeros(self.shape, dtype=self.dtype)
    result.ravel()[np.ravel_multi_index(self.coords, self.shape)] = self.data
```

anushkasuyal committed:
added trailing , in parametrize statements (following convention)

lucascolley commented:
cc @ilan-gold for visibility, as this may help towards array API compatibility

anushkasuyal committed:
changed np.prod to math.prod

anushkasuyal committed:
changed np.prod to math.prod

anushkasuyal committed:
modified ValueError for format conversions

anushkasuyal commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +576,90 @@ def _add_dense(self, other):
+            return result
+        
+        else: # if dim >= 3
+            result = np.zeros(np.prod(self.shape[:-1]),
```
Thanks! Just made the modification at 906c97b.

anushkasuyal replied:
(I accidentally undid the commit so a more recent commit shows a change back from `math.prod` to `np.prod`, but that's been fixed now!)

anushkasuyal commented on scipy/sparse/_coo.py:
```diff
@@ -318,6 +326,9 @@ def tocsc(self, copy=False):
         """
+        if self.ndim > 2:
+            raise ValueError('Cannot convert a higher dimension sparse array to '
+                             'csc format')
```
Thanks!
I noticed that removing the 'Cannot convert' was causing some tests (related to errors raised during format conversion) to fail, as I think they originally tried maintaining uniformity in the error messages for all conversions. So I kept the 'Cannot convert' at the beginning, as follows -
```
if self.ndim != 2:
            raise ValueError('Cannot convert.'
                             f"csc format sparse arrays must be 2D. Got {self.ndim}D")
```
(and similar for CSC, DIA, DOK.)

Addressed at 5a46c7a.

anushkasuyal committed:
coord to coords

anushkasuyal commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +576,90 @@ def _add_dense(self, other):
+            coo_todense(M, N, self.nnz, self.row, self.col, self.data,
+                        result.ravel('A'), fortran)
+        else:
+            coord = np.concatenate(self.coords)
```
Fixed at 83167de.

anushkasuyal committed:
renamed variables for better readability

anushkasuyal committed:
fixed bug in _matmul_dispatch and added test case

anushkasuyal commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +576,90 @@ def _add_dense(self, other):
+            return self._matmul_vector(other)
+        if other.shape == (N, 1):
+            result = self._matmul_vector(other.ravel())
+            return result.reshape(M, 1)
```
Ah, yes! Sorry, that was definitely a bug and I completely overlooked that case during testing due to which I couldn't catch the error then, so thanks a lot! I've addressed this now at 877faca, and added tests for this case too.

dschult commented on scipy/sparse/tests/test_coo.py:
```diff
@@ -268,6 +541,20 @@ def test_2d_matmul_multivector():
     assert_equal(res.toarray(), exp)
 
 
+@pytest.mark.parametrize(('mat_shape', 'vec_shape'), [((2,3,4,5), (5,)), ((0,0), (0,)),
+                                            ((2,3,4,7,8), (8,)), ((4,4,2,0), (0,)),
+                                            ((6,5,3,2,4), (4,1)), ((2,5), (5,1)), ((3,), (3,1)),])
```
These are hard to read, near the end of the line limit (88 chars), and not uniformly indented which would make it worse. Perhaps the most readable aproach is to create a variable on the previous line to hold all the cases.
```python
mat_vec_shapes = [
    ((2, 3, 4, 5), (5,)), 
    ((0, 0), (0,)), 
    ((2, 3, 4, 7, 8), (8,)),
    ((4, 4, 2, 0), (0,)),
    ((6, 5, 3, 2, 4), (4, 1)), 
    ((2, 5), (5, 1)),
    ((3,), (3, 1)),
]
```
followed by the parametrize decorator.  That will help the `lint` CI test to pass.

anushkasuyal replied:
Thanks - fixed at c3b9594.

dschult commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +572,90 @@ def _add_dense(self, other):
+        if self.ndim < 3:
+            return _data_matrix._matmul_dispatch(self, other)
+        
+        M, N = self.shape[-2], self.shape[-1]
```
linting CI test shows M not used any more.
```suggestion
        N = self.shape[-1]
```

anushkasuyal committed:
RandomState replaced with default_rng, random numpy arrays given a seed

anushkasuyal commented:
> The tests that use `randn` to construct random numpy arrays should either be made not random, or be given a seed to make them deterministic. Its hard to debug a test that only fails occasionally. Also, I think the tests with RandomState should (by numpy recommendation) be changed to use the random generator interface -- even though the keyword they are provided to is named `random_state`. :}

Addressed at 6862fce.

anushkasuyal committed:
Ac renamed to Aijk

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
formatting to pass lint check

rgommers commented:
Wow, this is moving fast! Hard to keep up:) Thanks @anushkasuyal, @lucascolley and @dschult!

> Also, I think the tests with RandomState should (by numpy recommendation) be changed to use the random generator interface

Small correction: this is actually not the case, `RandomState` is guaranteed to be more stable (since algorithms are frozen, unlike for `Generator`), and we're using it all over for testing. From [NEP 19](https://numpy.org/neps/nep-0019-rng-policy.html):
 _"The purpose of `RandomState` will be documented as providing certain fixed functionality for backwards compatibility and stable numbers for the limited purpose of unit testing."_ 

Using `default_rng` in testing is not really an issue, but when writing new code I'd prefer `RandomState` and when reviewing I'd accept both and not request changes.

anushkasuyal commented:
Thanks @rgommers! Shall I revert back the changes to `RandomState` in that case?

anushkasuyal committed:
casting coords to dtype int64 for dim>2 to prevent overflow in np.ravel_multi_index

rgommers commented:
> Shall I revert back the changes to `RandomState` in that case?

No need - fine to keep it as is now, since it's less work.

anushkasuyal committed:
Merge remote-tracking branch 'upstream/main'

anushkasuyal committed:
Merge branch 'main' into extend-coo-nd

anushkasuyal commented:
> And finally, I would really like to know how much slower it would be to use numpy tools instead of the `coo_todense_nd` C++ code. Something like:
> 
> ```python
>     result = np.zeros(self.shape, dtype=self.dtype)
>     result.ravel()[np.ravel_multi_index(self.coords, self.shape)] = self.data
> ```
(For reference, this has been answered at https://github.com/scipy/scipy/issues/21169#issuecomment-2236261662)

dschult commented on scipy/sparse/_coo.py:
```diff
@@ -286,13 +293,17 @@ def toarray(self, order=None, out=None):
+                        B.ravel('A'), fortran)
+            
+        else:
+            coords = np.concatenate(self.coords)
+            coo_todense_nd(self.shape, self.nnz, self.ndim,
+                           coords, self.data, B.ravel('A'), fortran)
```
I thought you found that the numpy method was better than the sparsetools for nd. Can you explain what your plans are for sparsetools vs numpy tools? I think we talked about it in a zoom call, but it'd be nice to have it in github comments -- and I'm not sure I understood correct the approach you plan to take.

anushkasuyal replied:
Hi! From what I recall, the approach we all seemed to like was that we will initially try to get full support for n-D sparse arrays in at least one format. Once we have that fully working, if time permits we could then go back and look into optimising / refactoring the code. We will only make minimal C++ additions, and only those that are truly necessary (for which the NumPy alternatives are too slow).

As for the sparsetools vs numpy tools plans -

- sparse-dense add uses `coo_todense_nd` from sparsetools.
- sparse-sparse add and sub use numpy tools.
- mat-vec multiplication (n-D sparse * 1-D dense) uses `coo_matvec_nd` from sparsetools.

For the methods that will possibly be in the follow-up PR -

- `tensordot` and `dot` are a combination of numpy tools and pre-existing CSR sparsetools. They are mapping two n-D sparse arrays to two 2-D ones, using 2-D CSR sparse-sparse multiplication, and mapping the product back to n-D. I've tested this approach and it seems to work fine.
- sparse-sparse `matmul` is currently work in progress. The approach in mind is to convert n-D COO arrays to block diagonal form (i.e. a 2-D COO array), use CSR matmul, then convert the product back to n-D.

We are yet to look into what will be more efficient for sparse-dense `dot`, `tensordot`, and `matmul`, (we might need to work with compiled code like we did for sparse-dense matvec and sparse-dense addition) but we are aiming to get the sparse-sparse operations for the three working properly first.

But, again, all of the above will be in a follow-up PR. This PR is necessary as it makes the required changes to allow n-D indices in the COO data structure.

anushkasuyal committed:
error check added in _matmul_dispatch

rgommers commented:
Hi @anushkasuyal, as we discussed, here is the method for measuring binary size differences:

```
$ git checkout extend-coo-nd
$ git merge main  # ensure we're up to date when comparing our branch with main
$ # ensure we start from a clean slate to pick up release build flags (note: ensure you don't have unsaved files, or you'll lose them!)
$ git clean -xdf 
$ unset CFLAGS && unset CXXFLAGS && unset FFLAGS  # only needed if you use conda compilers

$ python dev.py build --release
$ ls -l build/scipy/sparse/sparsetools/*.so
3512096 ... build/scipy/sparse/_csparsetools.cpython-310-darwin.so

$ git checkout main
$ python dev.py build  # this will be fast (rebuilds only the sparsetools extension)
$ ls -l build/scipy/sparse/sparsetools/*.so
3478096  ... build/scipy/sparse/_csparsetools.cpython-310-darwin.so
```
The difference is 34 kb, which is small (so changes in this PR are fine from that perspective).

anushkasuyal commented:
A tracker was started at https://github.com/scipy/scipy/issues/21169#issuecomment-2271007490, to keep track of all methods for COO arrays, and the progress of their implementation for n-D, at each stage (in each PR).

The extensions to n-D made so far in the first PR are as follows -

| **Methods**                              | **1- and 2-D** | **n-D** | **Implementation**                                      |
| ------------------------------------ | ---------- | --- | --------------------------------------------------- |
| `reshape`                              | Y          | Y   | all-D in python                                   |
| `count_nonzero`                        | Y          | Y   | all-D in python                                   |
| `transpose`                            | Y          | Y   | all-D in python                                   |
| `resize`                               | Y          |     | 1- and 2-D in python                               |
| `toarray`                              | Y          | Y   | 1- and 2-D use `coo_todense`; n-D uses `coo_todense_nd` |
| `tocsc`                                | Y          | \-  |                                                     |
| `tocsr`                                | Y          | \-  |                                                     |
| `tocoo`                                | Y          | Y   |                                                     |
| `todia`                                | Y          | \-  |                                                     |
| `todok`                                | Y          | \-  |                                                     |
| `diagonal`                             | Y          |     | in python                                           |
| `sum_duplicates`                       | Y          | Y   | all-D in python                                       |
| `eliminate_zeros`                      | Y          | Y   | all-D in python                                       |
| `_add_dense`                           | Y          | Y   | 1- and 2-D use `coo_todense`; n-D uses `coo_todense_nd` |
| `_add_sparse`                          | Y          | Y   | 1- and 2-D route via CSR; n-D uses python           |
| `_sub_dense`                           | Y          | Y   | all-D convert to dense and subtract                 |
| `_sub_sparse`                          | Y          | Y   | 1- and 2-D route via CSR; n-D uses python           |
| ` _matmul_vector` (sparse-dense)     | Y          | Y   | 1- and 2-D use `coo_matvec`; n-D uses `coo_matvec_nd`   |
| `_matmul_multivector` (sparse-dense) | Y          |     | 1- and 2-D use `coo_matmat_dense`                     |
| `_matmul_sparse` (sparse-sparse)     | Y          |     | 1- and 2-D route via CSR                            |
| `multiply`                             | Y          |     | via CSR                                             |
| `maximum`                              | Y          |     | via CSR                                             |
| `minimum`                              | Y          |     | via CSR                                             |
| `max`                                  | Y          |     |implemented upto 2-D in `_data_matrix`  |
| `min`                                  | Y          |     | implemented upto 2-D in `_data_matrix`  |
| `nanmax`                               | Y          |     | implemented upto 2-D in `_data_matrix`  |
| `nanmin`                               | Y          |     | implemented upto 2-D in `_data_matrix` |
| `power`                                | Y          | Y   | implemented for all-D in `_data_matrix`  |
| `eq`                                   | Y          |     | via CSR                                             |
| `ne`                                   | Y          |     | via CSR                                             |
| `lt`                                   | Y          |     | via CSR                                             |
| `gt`                                   | Y          |     | via CSR                                             |
| `le`                                   | Y          |     | via CSR                                             |
| `ge`                                   | Y          |     | via CSR                                             |
| `abs`                                  | Y          | Y   | uses built-ins in `_data_matrix`                     |
| `round`                                | Y          | Y   | uses np.around in `_data_matrix`                     |
| `div`                                  | Y          |     | uses numpy and routes via CSR                       |
| `dot` (sparse-dense)                   | Y          |     | 1- and 2-D uses matmul                              |
| `dot` (sparse-sparse)                 | Y          |     | 1- and 2-D uses matmul                              |
| `tensordot` (sparse-dense)             |            |     |                                                     |
| `tensordot` (sparse-sparse)            |            |     |                                                     |

anushkasuyal commented:
Additional note - in a later PR, we intend on removing `coo_todense_nd` and replacing its usage in `toarray()` with the NumPy definition (as provided by @dschult at https://github.com/scipy/scipy/issues/21169#issuecomment-2231275707). The 3rd PR will implement some linalg operations along with element-wise boolean comparisons and arithmetic operations (eq, gt, ne etc, and maximum, minimum, multiply etc.) by broadcasting and routing via CSR, and we will also refactor sparse-dense addition then to completely eliminate `coo_todense_nd`, such that it only relies on CSR `_add_dense`, as we extend all element-wise operations to use broadcasting, similar to NumPy.

rgommers committed:
Merge branch 'main' into extend-coo-nd

rgommers commented on scipy/sparse/tests/test_coo.py:
```diff
@@ -102,43 +167,99 @@ def test_reshape():
-
-def test_nnz():
-    arr1d = coo_array([1, 0, 3])
-    assert arr1d.shape == (3,)
```
Note that I'm looked at the unified diff with whitespace removed:

<img width="319" alt="image" src="https://github.com/user-attachments/assets/221c2afa-b164-40bb-969a-c5125e417883">

and also checking the whole file locally for lines that appear to be deleted in this diff view.

anushkasuyal replied:
I've attempted to move all tests parametrised for n-D to the bottom to maintain the ordering of tests in the original file. I've only made minimal additions to pre-existing tests (example, where a 2-D case was missing). Addressed at bee746c. If you'd prefer that I completely leave the original tests as they are, without any additions, I can do so!

rgommers replied:
This is great, thanks! I'll resolve this comment.

rgommers commented on scipy/sparse/sparsetools/coo.h:
```diff
@@ -112,6 +112,53 @@ void coo_todense(const I n_row,
+ *   T  Bx[]             - dense array
+ *
+ */
+template <class I, class T>
```
This is consistent with the other sparsetools code, but I'm just noticing it now because we're having persistent issues with using `I` as variable name (see for example gh-21376). Let's rename them all in a separate PR - we can touch on this today (not blocking for this PR).

rgommers commented on scipy/sparse/_sputils.py:
```diff
@@ -241,21 +241,33 @@ def isintlike(x) -> bool:
+        if not allow_nd:
             return False
-    return True
+        for d in x:
```
This block doesn't have to be duplicated from above. I think this function can be written as:
```python
    if ndim < 3:
        if ndim != 2 and not (allow_1d and ndim == 1):
            return False
    elif not allow_nd:
        return False
    else:
        for d in x:
            ...
```

anushkasuyal replied:
Addressed at 8ec8bd0.

rgommers commented on scipy/sparse/_sputils.py:
```diff
@@ -331,14 +349,25 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+                raise ValueError('shape must be a 2-tuple of positive integers')
+            if any(d < 0 for d in new_shape):
+                raise ValueError("'shape' elements cannot be negative")
+        else: # dim >= 3
```
This block is largely duplicated as well, can you try to reduce that (similar to my comment for `isshape`)?

rgommers commented on scipy/sparse/_sputils.py:
```diff
@@ -362,8 +391,14 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+    if len(new_shape) < 3:
+            if len(new_shape) != 2 and not (allow_1d and len(new_shape) == 1):
+                raise ValueError('matrix shape must be two-dimensional')
+    elif len(new_shape) >= 3:
```
Some duplication here too, I suggest:
- define `ndim_new = len(new_shape)` and use `ndim_new` in the checks
- try to unify the checks that yield an identical error message (`'matrix shape must be two-dimensional'` in this case)

anushkasuyal replied:
Addressed at 8ec8bd0.

rgommers reviewed (commented):
This looks great and is very close, thanks @anushkasuyal! Some comments:

Some of the tests looked a bit overparametrized at first sight, but on closer inspection the choices do serve separate purpose - size 0, size 1, and some other special cases seem to be well-chosen. The test runtime increase is very small (`35 passed in 0.34s` becomes `154 passed in 0.35s`), which is nice.

I noticed that `coo_array.todense` works (as far as I tried it), but isn't explicitly tested or added to your table of methods yet. Could you add that? Especially relevant because you're adding new C++ code for it here.

Regarding the added `allow_nd` boolean: would it make sense to have that always imply `allow_1d`? I think there shouldn't be cases where  n-D is allowed but 1-D is not. If that is the case, check and code like this can be simplified (example from `_coo_base`):
```python
check_shape(shape, allow_1d=is_array, allow_nd=(is_array and len(shape)>=3))
```
It would be nicer to have code at the top of a function like
```python
allow_1d=is_array
allow_nd= is_array and len(shape)>=3
# and then everywhere else:
check_shape(shape, allow_1d=allow_1d, allow_nd=allow_nd)
```
and inside `check_shape` there's no need to check for `allow_1d` in code blocks that handle n-D.

rgommers commented on scipy/sparse/_sputils.py:
```diff
@@ -331,14 +349,25 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+                if any(d < 0 for d in new_shape):
+                    raise ValueError("'shape' elements cannot be negative")
+            else:
+                if allow_1d:
```
This is an example of a check using `allow_1d` inside a >=3 block - can be removed.

anushkasuyal replied:
Modified at 8ec8bd0. Is the current structure alright or would you advise I further unify the conditions for the `'shape must be a 2-tuple of positive integers'` error?

lucascolley commented on scipy/sparse/tests/test_coo.py:
```diff
@@ -171,7 +292,7 @@ def test_1d_toformats():
 
 
-@pytest.mark.parametrize('arg', [1, 2, 4, 5, 8])
+@pytest.mark.parametrize('arg', [1, 2, 4, 5, 8,])
```
yeah, this might be my fault for not being more specific in an earlier review. [PEP8 says](https://peps.python.org/pep-0008/#when-to-use-trailing-commas):

> When trailing commas are redundant, they are often helpful when a version control system is used, when a list of values, arguments or imported items is expected to be extended over time. The pattern is to put each value (etc.) on a line by itself, always adding a trailing comma, and add the close parenthesis/bracket/brace on the next line. However it does not make sense to have a trailing comma on the same line as the closing delimiter (except in the above case of singleton tuples):

so they should be avoided for lists that fit on a single line

anushkasuyal committed:
old tests added back, tests reordered to minimize diff

anushkasuyal committed:
changes for readability

anushkasuyal committed:
duplications removed in isshape and checkshape

anushkasuyal committed:
todense tests added

anushkasuyal committed:
random_array tests added

anushkasuyal committed:
remove test cases for shape check of random_array

anushkasuyal commented on scipy/sparse/tests/test_coo.py:
```diff
@@ -102,43 +167,99 @@ def test_reshape():
+    rng = np.random.default_rng(23409823)
+
+    arr = random_array(shape, density=0.6, random_state=rng, dtype=int)
+    assert arr.shape == (shape)
```
I've added `todense`/`hstack`/`vstack`/`block_array`/`random_array` to the feature tracking table. Out of these, `todense` and `random_array` are currently functional for n-D. I've added tests in `test_construct.py` for `random_array`, at d2cb7a1. The other three can then be covered in a future PR.

I've also removed the above test cases from `test_coo.py`, at 733a78b.

rgommers replied:
That all looks good, thanks - resolving.

anushkasuyal commented:
Thanks @rgommers ! Addressing a few points below -

> I noticed that `coo_array.todense` works (as far as I tried it), but isn't explicitly tested or added to your table of methods yet. Could you add that? Especially relevant because you're adding new C++ code for it here.

I've added `todense` to the table of methods, and added tests for it at 023954b.

> Regarding the added `allow_nd` boolean: would it make sense to have that always imply `allow_1d`? I think there shouldn't be cases where n-D is allowed but 1-D is not.

Yes, I agree that there shouldn't be such a case. I don't think any methods implemented so far are such that they are defined for n-D, but not for 1-D.

> If that is the case, check and code like this can be simplified (example from `_coo_base`):
> 
> ```python
> check_shape(shape, allow_1d=is_array, allow_nd=(is_array and len(shape)>=3))
> ```
> 
> It would be nicer to have code at the top of a function like
> 
> ```python
> allow_1d=is_array
> allow_nd= is_array and len(shape)>=3
> # and then everywhere else:
> check_shape(shape, allow_1d=allow_1d, allow_nd=allow_nd)
> ```
> 
> and inside `check_shape` there's no need to check for `allow_1d` in code blocks that handle n-D.

For this, I tried defining `allow_nd` as `is_array and len(shape)>=3` at the beginning of `__init__` but that didn't work, because on initialisation the array may or may not have a `shape` attribute. Also, I originally defined `allow_nd` as `allow_nd=is_array` or `allow_nd=is_array and len(shape)>=3` but I think `is_array` should be sufficient by itself, because all methods in `_coo.py` that call `check_shape` apart from `resize` have been implemented in n-D in this PR, and `resize` has been let to default `allow_nd` to `False` in the `check_shape` call.

Does it then sound sensible to change `allow_nd` to `is_array`?

rgommers commented on scipy/sparse/_sputils.py:
```diff
@@ -258,6 +262,7 @@ def isshape(x, nonneg=False, *, allow_1d=False) -> bool:
     return True
 
 
+
```
nit: 2 blank lines is enough, this is the 3rd one

anushkasuyal replied:
Removed at https://github.com/scipy/scipy/pull/21197/commits/89ffb7f50f2fb22ea3a33aa5e0c3f6ad29ca4bb3.

rgommers commented on scipy/sparse/_sputils.py:
```diff
@@ -297,7 +302,8 @@ def validateaxis(axis) -> None:
 
-def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+def check_shape(args, current_shape=None, *, allow_1d=False,
+                allow_nd=False) -> tuple[int, ...]:
```
minor: please feel free to delete ` -> tuple[int, ...]:`. Having only half a type annotation in a signature is not right.

rgommers commented on scipy/sparse/_sputils.py:
```diff
@@ -362,9 +376,13 @@ def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
         raise ValueError('matrix shape must be two-dimensional')
 
+    if ndim_new >= 3 and allow_1d and not allow_nd:
+        raise ValueError('matrix shape must be one- or two-dimensional')
```
While we're at it, can you remove the word `matrix` from the error message (also 2 lines up)? This was pre-existing, but we have arrays now and not only matrices!

anushkasuyal replied:
Changed at https://github.com/scipy/scipy/pull/21197/commits/839d462fb775caac9cffc3a842824a7bdd54a514.

rgommers reviewed (commented):
Nice work @anushkasuyal! All comments from my previous review are addressed, and this is basically ready to merge I think, modulo the few small things I just commented on.

This is now very readable and concise, and the added test coverage without changing anything for the 1-D/2-D tests gives me confidence that this is fully backwards compatible.

@dschult @perimosocordiae I'm hoping to revisit and merge this PR next week, during EuroSciPy next Wed-Fri.  Are you planning to review this or have any concerns with that plan?

anushkasuyal committed:
allow_nd assignment changed

anushkasuyal committed:
extra brackets removed

anushkasuyal committed:
extra blank line removed

anushkasuyal committed:
TYPO: 'matrix' removed

anushkasuyal commented:
@rgommers all conversations have been resolved now, thanks!

dschult commented:
@anushkasuyal I think one conversation (function signature type annotation) still remains.

@rgommers That timing for further review and possible merging later in next week sounds good to me. I should have some time to look at it again this coming weekend.

anushkasuyal commented on scipy/sparse/_sputils.py:
```diff
@@ -297,7 +301,8 @@ def validateaxis(axis) -> None:
 
-def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+def check_shape(args, current_shape=None, *, allow_1d=False,
+                allow_nd=False) -> tuple[int, ...]:
```
```suggestion
                allow_nd=False)
```

anushkasuyal committed:
Update _sputils.py

anushkasuyal commented:
Ah yes, sorry I missed that one! Fixed it now, thanks @dshult

dschult commented on scipy/sparse/_sputils.py:
```diff
@@ -297,7 +301,8 @@ def validateaxis(axis) -> None:
 
 
-def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+def check_shape(args, current_shape=None, *, allow_1d=False,
+                allow_nd=False)
```
Can we combine the parameters `allow_1d` and `allow_nd`? I believe they always take on the value `is_array` which just checks for sparray (and thus not spmatrix). As far as shapes are concerned, sparrays should allow nD. Specific formats may not support nD, but that is checked elsewhere -- no check of the shape should depend on the format. 

Some background: When we added support for 1D arrays we first wrote it using the parameter name `allow_nd`. We changed it to `allow_1d` when we delayed support for nd.

There doesn't seem to be any need for two parameters in `isshape` and in `check_shape`. The long term solution is to have `allow_nd` carry this information.  I would be fine with making the switch. But if people want a deprecation cycle, that should work too. Keep the old `allow_1d`, add the `allow_nd` and at the beginning of the functions check whether `allow_1d` is not its default, in which case raise a DeprecationWarning and set  `allow_nd` to the value given. Otherwise just use `allow_nd`. After two releases the deprecation can be backed out.

Do you have thoughts about using both parameters over just using `allow_nd`?

anushkasuyal replied:
The primary concern for having both separately is because other formats would allow 1d but now allow nd, whereas for COO we would allow both. This would not break any existing 1D stuff for other formats, and if someone in the future wants to extend other formats such as CSR to n-D, that can be done easily by setting `allow_nd` to `is_array` instead of letting it default to `False`.
If we were to combine `allow_1d` and `allow_nd`, I'm a bit concerned about the handling of, example, the following code from [_sputils.py](https://github.com/anushkasuyal/scipy/blob/f36e52d28582007619d9717cd9f8924694447118/scipy/sparse/_sputils.py#L343-L351) -
```py
    if current_shape is None:
        if not allow_1d and len(new_shape) != 2:
            raise ValueError('shape must be a 2-tuple of positive integers')

        if not allow_nd and len(new_shape) > 2:
            if allow_1d:
                raise ValueError('shape must be a 1- or 2-tuple of positive integers')
            else:
                raise ValueError('shape must be a 2-tuple of positive integers')
```
We do want to raise the "shape must be a 1- or 2-tuple of positive integers" error for other formats, (which allow 1d but not nD), but for COO there is no such case.

anushkasuyal replied:
`allow_1d` and `allow_nd` have now been combined into a tuple-valued parameter, `allowed_ndims`, at https://github.com/scipy/scipy/commit/0ea748b03fb949919287bc9f57ba0e0df0fdad4a.

dschult commented on scipy/sparse/_coo.py:
```diff
@@ -423,6 +439,9 @@ def todia(self, copy=False):
 
     def todok(self, copy=False):
+        if self.ndim > 2:
+            raise ValueError('Cannot convert. dok format sparse arrays must be'
+                             f" 2D or 1D. Got {self.ndim}D")
```
```suggestion
            raise ValueError(f'Cannot convert. DOK must be 1D or 2D. Got {self.ndim}D')
```

dschult commented on scipy/sparse/_coo.py:
```diff
@@ -400,7 +415,8 @@ def tocoo(self, copy=False):
     def todia(self, copy=False):
         if self.ndim != 2:
-            raise ValueError("Cannot convert a 1d sparse array to dia format")
+            raise ValueError('Cannot convert.'
+                             f"dia format sparse arrays must be 2D. Got {self.ndim}D")
```
```suggestion
            raise ValueError(f'Cannot convert. DIA format must be 2D. Got {self.ndim}D')
```

dschult commented on scipy/sparse/_coo.py:
```diff
@@ -351,6 +363,9 @@ def tocsr(self, copy=False):
 
         """
+        if self.ndim > 2:
+            raise ValueError('Cannot convert.'
+                             f"csr format sparse arrays must be 2D. Got {self.ndim}D")
```
```suggestion
            raise ValueError(f'Cannot convert. CSR must be 1D or 2D. Got {self.ndim}D')
```

dschult commented on scipy/sparse/_coo.py:
```diff
@@ -319,7 +330,8 @@ def tocsc(self, copy=False):
         """
         if self.ndim != 2:
-            raise ValueError("Cannot convert a 1d sparse array to csc format")
+            raise ValueError('Cannot convert.'
+                             f"csc format sparse arrays must be 2D. Got {self.ndim}D")
```
```suggestion
            raise ValueError(f'Cannot convert. CSC format must be 2D. Got {self.ndim}D')
```

dschult commented on scipy/sparse/_sputils.py:
```diff
@@ -241,14 +241,18 @@ def isintlike(x) -> bool:
     """
     ndim = len(x)
-    if ndim != 2 and not (allow_1d and ndim == 1):
+    if ndim < 3:
+        if ndim != 2 and not (allow_1d and ndim == 1):
+            return False
+    elif not allow_nd:
```
With just the allow_nd parameter this code could be simplified to something like:
```python
    if not allow_nd and ndim != 2:
        return False
```

dschult reviewed (commented):
This looks good. 

I would prefer to replace `allow_1d` with `allow_nd` rather than keeping both. But I understand that deprecations might be needed. 

I also made some nit-picky suggestions for ndim related error messages. The DOK and CSR formats support 1D and 2D now, so CSR should be changed to reflect that. The other suggestions remove confusion about sparse arrays vs sparse matrices and they shorten the message so it appears on one line.

anushkasuyal committed:
missing colon

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

perimosocordiae commented on scipy/sparse/sparsetools/coo.h:
```diff
@@ -112,6 +112,53 @@ void coo_todense(const I n_row,
+ *
+ */
+template <class I, class T>
+void coo_todense_nd(const I shape[],
```
Instead of passing in a `shape[]` array, how about passing in the cumulative product as a `strides[]` array instead? That would avoid having to re-calculate the stride every iteration of the main loop.

anushkasuyal replied:
Just to confirm, we would need to pass both C and Fortran strides for `coo_todense_nd`, right?
Addressed at d56b75ef1728004e53657da892d73ad6bf555eee.

dschult replied:
We don't actually need both C and F strides. If we look at the value we input for `fortran` we can figure out which strides to input.

anushkasuyal replied:
Thanks, modified at f9ba32125b0a286f2e50a405263d40eded6465a0.

perimosocordiae commented on scipy/sparse/sparsetools/coo.h:
```diff
@@ -146,6 +193,40 @@ void coo_matvec(const npy_int64 nnz,
+template <class I, class T>
+void coo_matvec_nd(const npy_int64 nnz,
+                const npy_int64 num_dims,
+                const I shape[],
```
Same comment here: we could pass `strides[]` instead of `shape[]`.

anushkasuyal replied:
Addressed at 54d60d6.

perimosocordiae commented on scipy/sparse/_sputils.py:
```diff
@@ -297,7 +301,8 @@ def validateaxis(axis) -> None:
 
 
-def check_shape(args, current_shape=None, *, allow_1d=False) -> tuple[int, ...]:
+def check_shape(args, current_shape=None, *, allow_1d=False,
```
Here's another option for combining `allow_1d` with `allow_nd`: pass the keyword argument `allowed_ndims=(2,)` to represent the set of valid dimensionalities.

So by default, we'd have (2,) for only 2d shapes, but CSR and DOK could pass `(1, 2)` and COO could pass `range(1, 65)` (or some large upper limit).

This should hopefully simplify the checking logic, and also makes it explicit which values are allowed at the call-site.

anushkasuyal replied:
Addressed at 0ea748b. Docstrings edited at https://github.com/scipy/scipy/pull/21197/commits/e568d568647deac8aa65d751b0424bcb7822f316.

perimosocordiae commented on scipy/sparse/_coo.py:
```diff
@@ -286,13 +293,17 @@ def toarray(self, order=None, out=None):
-        M, N = self._shape_as_2d
-        coo_todense(M, N, self.nnz, self.row, self.col, self.data,
-                    B.ravel('A'), fortran)
+        if self.ndim < 3:
```
Let's call `coo_todense_nd` for the 1-dimensional case as well, because otherwise we have to allocate a temporary array of all zeros when we access `self.row`.

anushkasuyal replied:
Addressed at 69e8757.

perimosocordiae commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +571,94 @@ def _add_dense(self, other):
-        M, N = self._shape_as_2d
-        coo_todense(M, N, self.nnz, self.row, self.col, self.data,
-                    result.ravel('A'), fortran)
+        if self.ndim < 3:
```
Same here, let's use the nd version for 1-dimension cases.

anushkasuyal replied:
Addressed at 69e8757.

perimosocordiae commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +571,94 @@ def _add_dense(self, other):
-                          dtype=upcast_char(self.dtype.char, other.dtype.char))
+    def _add_sparse(self, other):
+        if self.ndim < 3:
+            return _data_matrix._add_sparse(self, other)
```
`_data_matrix` doesn't define `_add_sparse()`, but its parent class `_spbase` does. And that implementation is converting to CSR format.

Let's do the CSR conversion explicitly here. (and in `_sub_sparse`).

anushkasuyal replied:
I've both added the CSR conversion and changed the class we are calling `_add_sparse` from. Addressed at https://github.com/scipy/scipy/pull/21197/commits/3f7278228a79c9e47eefce3459bd065b8847e731.

dschult replied:
I think @perimosocordiae intended you to inline the one-line code from `_spbase` for both `_add_sparse` and `_sub_sparse`. As currently changed, the `tocsr()` method would get called twice. Instead use something like this:
```python
        if self.ndim < 3:
            return self.tocsr()._add_sparse(other)
```
This way the reader doesn't even have to look up the method (and for a one line method like this it is easier to read this line than to look through the inheritance.)

anushkasuyal replied:
Modified at 1322cb5179982afb46a1df30d19def66b653d7a7.

perimosocordiae commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +571,94 @@ def _add_dense(self, other):
+                return result[0]
+            return result
+        
+        else: # if dim >= 3
```
Let's move this condition to the top, and then because we unconditionally return, we won't need to indent the existing code (to make the diff a bit cleaner).

anushkasuyal replied:
Addressed at https://github.com/scipy/scipy/pull/21197/commits/46e18b70301c944d56cf8c68aac6209be6f73291.

perimosocordiae commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +571,94 @@ def _add_dense(self, other):
+
+    def _matmul_dispatch(self, other):
+        if self.ndim < 3:
+            return _data_matrix._matmul_dispatch(self, other)
```
Let's use `_spbase._matmul_dispatch()` here for clarity.

anushkasuyal replied:
https://github.com/scipy/scipy/pull/21197/commits/545ad3fb275792f2b6f0727bb261ad5e0a255777.

perimosocordiae commented on scipy/sparse/_coo.py:
```diff
@@ -556,31 +571,94 @@ def _add_dense(self, other):
+            return _data_matrix._matmul_dispatch(self, other)
+        
+        N = self.shape[-1]
+        if other.shape == (N,):
```
We need to check that `other` isn't sparse before dispatching here, because 3d-sparse @ 1d-sparse won't work in this code path.

anushkasuyal replied:
Ah yes, I missed that out. Also, I noticed we called `_spbase._matmul_dispatch` if `self.ndim<3`, irrespective of `other`'s dimension. This does throw an error at some point if we multiply, example `2-D * 5-D`, because we don't have n-D `matmul` yet, but it will eventually cause issues when we add `matmul` in the next PR, so I just extended the conditions checked by the if statement. Do the fixes at e6004d3ad2a46dc00ebc38ad40791a6750ae7865 work fine?

dschult replied:
I think we want to avoid the computation used here to figure out whether to use the existing 2D matmul, especially the `asarray(other)` that is only used to get its shape. 
Furthermore, I think you shouldn't try to cover the case of `other` being sparse with large ndim here. The `_spbase._matmul_dispatch` call will return `NotImplemented` and Python will try `__rmatmul__` which will call `_rmatmul_dispatch` which switches the arguments using transposes and then calls this with the shapes changed.   So, I think the initial check can just be 
```python
if self.ndim < 3:
    return _spbase._matmul(self, other)
```

I think I don't understand the approach you are using for the rest of the `_matmul_dispatch` code here. It looks very similar to the 2D version but stops before handling any array-like (usually a list-of-lists). Can you briefly say here how this differs from the 2D approach? for example, are we removing the ability to matmul with a 2D list of lists?

anushkasuyal replied:
I'll remove the `asarray(other)` computation and revert the if condition back to what it was.

As of the array-like cases, I originally planned on handling that altogether when all n-D cases of matmul were working, but I think it may be wiser to have matrix-vector multiplication fully working in this PR, so I will incorporate those cases too.

While exploring the current 2-D mat-vec matmul, I noticed the following variations in behaviour -
```py
>>> a = coo_array([1,2,3])
>>> b = coo_array([[4],[5],[6]], shape=(3,1))
>>> dense = a.toarray() @ b.toarray()
>>> sparse_dense = a @ b.toarray()
>>> sparse_denselist = a @ list(b.toarray())
>>> sparse_sparse = (a@b).toarray()
>>> dense
array([32])
>>> sparse_sparse
array([32])
>>> sparse_dense
np.int64(32)
>>> sparse_denselist
array([[32]])
```

Was this intentional, or is this something I should fix for existing 2-D matmul, so that it is consistent with `np.matmul`'s behaviour?

dschult replied:
The rules we intend to support for matmul are the [array_api standards for matmul](https://data-apis.org/array-api/2023.12/API_specification/generated/array_api.matmul.html). There are two descriptions on that page (one about handling input shapes and the other handling output shapes).

It might make sense to split changes to existing 1/2D code into a separate PR, if only to keep them small. But if that makes it more complicated to review than a single PR go with a single PR. Review time seems to be the restricting features of code maintenance and upkeep.

One difficulty with this part of the code is that it was originally written for 2D systems so "vec" meant a 2D row or column. It is now tempting to think of a 1D object as a "vec" -- and that often works, but not always. The best approach of course is to do what you did in this comment and ignore the method names -- instead try the top level expressions to see if they mimic what numpy/array_api makes happen.

anushkasuyal replied:
I've tested the top level expressions and they pass the matmul tests against numpy's behaviour now. If you'd prefer, I can make a separate PR for the 1/2D code change.
The only issue with that might be that it would take time to be reviewed, hence blocking this conversation from being resolved.
It's only the following if statement added, so if it's easier to review it in a commit in this PR itself, that also works -

Instead of
```py
if other.ndim == 1 or other.ndim == 2 and other.shape[1] == 1:
   ...
   result = self._matmul_vector(np.ravel(other))
   ...
   if other.ndim == 2 and other.shape[1] == 1:
       # If 'other' was an (nx1) column vector, reshape the result
       result = result.reshape(-1, 1)
    return result
```

we would have
```py
if other.ndim == 1 or other.ndim == 2 and other.shape[1] == 1:
    ...
    result = self._matmul_vector(np.ravel(other))
    ...
    if other.ndim == 2 and other.shape[1] == 1:
        # If 'other' was an (nx1) column vector, reshape the result
        if self.ndim == 1:
            result = result.reshape(-1)
        else:
            result = result.reshape(-1, 1)
    return result

```

(and a similar one liner in another part of the method).

dschult replied:
I think a PR with those two one-liners and tests that show the undesired behavior has been corrected will be quickly reviewed and merged.  I don't think that would block this conversation from being resolved... It's a separate topic from needing to check whether `other` is sparse before dispatching. I guess there could be conflicts created by one PR when the other PR is merged, but they should be easy to resolve -- there are only two lines that could conflict.

I also don't think resolving the current 1-2D shape issues will stop this PR from being reviewed/merged. That is an existing problem, not a new feature.

Note that these are only my suggestions/perspectives based on my experience. They are not factual statements. You can put the one-liners in where you think they best fit. They will fix current bugs.

anushkasuyal committed:
coo_todense_nd used for 1-d cases

anushkasuyal committed:
explicit conversion to CSR in _add_sparse, _sub_sparse

anushkasuyal committed:
if-condition order switched for cleaner diff

anushkasuyal committed:
_matmul_dispatch called from _spbase

anushkasuyal committed:
_matmul_dispatch modified, checks for sparse/dense other added

anushkasuyal committed:
strides instead of shape argument in coo_todense_nd

anushkasuyal committed:
shape argument replaced with strides in coo_matvec_nd

anushkasuyal committed:
allow_1d and allow_nd replaced by allowed_ndims

anushkasuyal committed:
lint errors fixed

anushkasuyal committed:
docstrings updated to include allowed_ndims
[docs only]

anushkasuyal committed:
modified coo_todense_nd to have one strides parameter

dschult commented on scipy/sparse/_coo.py:
```diff
@@ -556,16 +577,68 @@ def _add_dense(self, other):
         result = np.zeros(result_shape,
-                          dtype=upcast_char(self.dtype.char, other.dtype.char))
-
+                            dtype=upcast_char(self.dtype.char, other.dtype.char))
```
```suggestion
                          dtype=upcast_char(self.dtype.char, other.dtype.char))
```

anushkasuyal committed:
explicit csr conversion for add, sub

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

anushkasuyal committed:
n-D _matmul_dispatch extended for array-like dense inputs

anushkasuyal committed:
pull changes

anushkasuyal committed:
typo

dschult commented on scipy/sparse/_coo.py:
```diff
@@ -286,13 +291,25 @@ def toarray(self, order=None, out=None):
+            M, N = self._shape_as_2d
+            coo_todense(M, N, self.nnz, self.row, self.col, self.data,
+                        B.ravel('A'), fortran)
+            
+        else:
+            shape = np.array(self.shape)
+            if fortran:
+                strides = np.append(1, np.cumprod(shape[:-1]))
+            else:
+                strides = np.append(np.cumprod(shape[1:][::-1])[::-1], 1)
```
A very minor tweak:  remove blank line and we can use the `self.shape` tuple to make `stride` to avoid a conversion.
```suggestion
        else:
            if fortran:
                strides = np.append(1, np.cumprod(self.shape[:-1]))
            else:
                strides = np.append(np.cumprod(self.shape[1:][::-1])[::-1], 1)
```

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

dschult commented on scipy/sparse/_coo.py:
```diff
@@ -556,16 +577,68 @@ def _add_dense(self, other):
+            M, N = self._shape_as_2d
+            coo_todense(M, N, self.nnz, self.row, self.col, self.data,
+                        result.ravel('A'), fortran)
+        else:
+            shape = np.array(self.shape)
+            if fortran:
+                strides = np.append(1, np.cumprod(shape[:-1]))
+            else:
+                strides = np.append(np.cumprod(shape[1:][::-1])[::-1], 1)
```
I guess this is the same as above....   
```suggestion
        else:
            if fortran:
                strides = np.append(1, np.cumprod(self.shape[:-1]))
            else:
                strides = np.append(np.cumprod(self.shape[1:][::-1])[::-1], 1)
```

Also, do you want me to add a commit to fix the linting (which is not part of what you've touched, but is close enough that ruff is checking and making suggestions...). Once I push it you would need to pull your branch to any local repo before making more changes.

anushkasuyal replied:
Yes, that sounds good, thanks!

anushkasuyal committed:
Update scipy/sparse/_coo.py

Co-authored-by: Dan Schult <dschult@colgate.edu>

dschult committed:
fix nearby linting issues due to type annotations

dschult committed:
undo broken linting fix

dschult reviewed (approved):
I think this is ready for merging. I believe all of @perimosocordiae comments are addressed, though they should look at them too. @rgommers said this might go in during EuroScipy, and I think it's ready for that.
Thanks @anushkasuyal for all this work!

dschult committed:
switch from allow_1d to allow_nd

dschult committed:
remove spaces at end of line

anushkasuyal committed:
Update scipy/sparse/_coo.py

anushkasuyal committed:
Merge pull request #3 from dschult/extend-coo-nd

Proposed update to the extend-coo-nd branch

dschult commented on scipy/sparse/_compressed.py:
```diff
@@ -87,25 +86,24 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False, *, maxprint=None):
                 raise ValueError('unable to infer matrix dimensions') from e
 
-            self._shape = check_shape(self._swap((major_d, minor_d)), allow_1d=is_array)
+            self._shape = check_shape(self._swap((M, N)), allow_nd=self._allow_nd)
+            self._shape = check_shape(self._swap((M, N)), allow_nd=self._allow_nd)
```
This duplication snuck in somehow.
```suggestion
            self._shape = check_shape(self._swap((M, N)), allow_nd=self._allow_nd)
```

dschult commented on scipy/sparse/_coo.py:
```diff
@@ -286,13 +289,23 @@ def toarray(self, order=None, out=None):
+            coo_todense_nd(np.array([1]), self.nnz, self.ndim,
+                           self.coords[0], self.data, B.ravel('A'), fortran)
+        elif self.ndim == 2:
+            M, N = self._shape_as_2d
```
We now know that this is 2d, so use `self.shape` directly
```suggestion
            M, N = self.shape
```

dschult committed:
Merge branch 'main' into extend-coo-nd

dschult reviewed (approved):
This review restates my approval for this PR. The recent changes to store a class attribute `_allow_nd` which holds a tuple of allowed ndim values simplify code and centralize creation of these values to ease future maintenance.

dschult committed:
Apply suggestions from code review

dschult committed:
lint fix

dschult committed:
ignore mypy errors for _allow_nd in _coo and _csr

rgommers commented:
> I think this is ready for merging. I believe all of @perimosocordiae comments are addressed, though they should look at them too. @rgommers said this might go in during EuroScipy, and I think it's ready for that.

Hey @dschult, I unfortunately have been unable to work on SciPy for a few weeks, and that may last a while longer (see my email on the core dev mailing list). I think here you were already happy and only waiting for me to re-review and hit the green button. Since I'm unable to re-review, could I ask if you are able to merge this PR and help @anushkasuyal's next PR along as well? @izaid has done the bulk of the initial reviews and I think with both of you happy (and me too from what I remember last time I tested this PR), I think we can be pretty confident in Anushka's work; I hope it doesn't get blocked too much by my absence.

dschult commented:
I'm merging this which will make the other PRs (#21169 and #21435) need some rebasing/merging.
Thanks very much @anushkasuyal and @izaid !
